### PR TITLE
stevenblack-blocklist: rev bump and improvements of module + package

### PR DESCRIPTION
**Pull request by Frontear**
## Description of changes

- Wrapped the package with `stdenvNoCC.mkDerivation`
- Added a `passthru.updateScript` for @r-ryantm (tested with `nix-shell maintainers/scripts/update.nix --argstr package stevenblack-blocklist`
- Moved from `pkgs/tools/networking/stevenblack-blocklist` to `pkgs/by-name/st/stevenblack-blocklist`
- Changed `$out` output to contain the various hosts only, removing all else
- Added 5 distinct outputs `[ "ads" "fakenews" "gambling" "porn" "social" ]` that expose isolated host files as outputs.
- Changed the nixos module to use these new updated outputs (tested in a `nixos-generate-config` config using `nixos-rebuild build-vm`)

New outputs help reduce closure size by having significantly smaller outputs depending on isolated usecases. Before, the closure size was roughly 100MB, but with these changes even the full combination of closures will be roughly 20MB, a significantly better size. The original directory is left as a fallback with all the extraneous stuff removed, as those are likely not of interest to a user who simply wants hosts.

Functionality remains exactly the same as before, with the added bonus of smaller closure sizes overall.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Frontear** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>stevenblack-blocklist</li>
  </ul>
</details>

**Review by JohnRTitor — COMMENTED** [review]
Why is this not using something like `stdenvNoCC`? and it would be a good idea to add a `passthru.updateScript`, so bots like ryantm can automatically bump this.

And would be nice to get this moved to `pkgs/by-name`

**Comment by Frontear** [comment]
Honestly all great points. Ill move it and have the stdenv use noCC, as well as setup the bot to auto update this. One less PR, woe my contributions

**Comment by Frontear** [comment]
Result of `nixpkgs-review pr 321649` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>stevenblack-blocklist</li>
  </ul>
</details>

**Review by itslychee — APPROVED** [review]

**Review by JohnRTitor — APPROVED** [review]
Sweet

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/st/stevenblack-blocklist/package.nix:17]:
```suggestion

  dontConfigure = true;
  dontBuild = true;

```

**Review by JohnRTitor — COMMENTED** [review]
Squash stdenvNoCC and updateScript commit

**Comment by Frontear** [comment]
> Squash stdenvNoCC and updateScript commit

Done, and added `dont{Configure,Build} = true` in this commit as well.

**Review by JohnRTitor — APPROVED** [review]
Approved, checked 

```bash
tree $(nix build github:nixos/nixpkgs/refs/pull/321649/merge#stevenblack-blocklist --print-out-paths)
```

**Comment by JohnRTitor** [comment]
I believe, these readme.md's can be removed before copying?

image

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/st/stevenblack-blocklist/package.nix:25]:
```suggestion
    find . -type f -iname "readme.md" -exec rm {} +
    cp -r hosts alternates $out
```

Comment (Frontear) [reply pkgs/by-name/st/stevenblack-blocklist/package.nix:25]: Nice and simple change, I like it!

**Review by moni-dz — APPROVED** [review]

**Comment by itslychee** [comment]
I think that's a good idea!

**Comment by Frontear** [comment]
One small change, because theres no sane way to sanitize blockList as an override value, I instead think ill have four values

```nix
fakenews ? false
block ? false
...
```
Course this changes implementation slightly, but end result of closure size reduction will still be achieved, with the added bonus of no possible breakage from a bad override

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (JohnRTitor) [reply pkgs/by-name/st/stevenblack-blocklist/package.nix:25]: Within the squashed.

**Comment by Frontear** [comment]
A fair critique I recieved on the unofficial NixOS discord is that this doesn't reduce the amount of "work" done (still needs to clone full repo, so network usage and storage space remains high still, this ONLY benefits closure size).

A solution I thought up for this (though it may be very unconventional), was to pull only the specific hosts file via a `fetchurl`. This would look something like this

```nix
src = fetchurl {
  url = https://raw.githubusercontent.com/StevenBlack/hosts/${finalAttrs.version}/path/to/hosts;
  hash = ""; # fill in
};
```

I believe this is quite unconventional, but it could solve both the unnecessary work, and reduce closure size all in one. This alongside with the module changes could result in a significantly improved package + module.

Again, open to critiques and suggestions!

**Comment by itslychee** [comment]
All the users of this package are going to use all parts of this package at some point so I wouldn't concern yourself with the download sizes for Hydra, but rather worry about splitting the package up into outputs so users can use relevant parts for their usecases instead of downloading the entire `src`, and the host files seems to be where the big file sizes are coming from anyways, seemingly averaging around 3MBs per host file.

**Comment by Frontear** [comment]
> All the users of this package are going to use all parts of this package at some point so I wouldn't concern yourself with the download sizes for Hydra, but rather worry about splitting the package up into outputs so users can use relevant parts for their usecases instead of downloading the entire `src`, and the host files seems to be where the big file sizes are coming from anyways, seemingly averaging around 3MBs per host file.

Honestly good point, I also don't particularly like the fetchurl thing. Backtracking on that idea, instead only focusing on simply opening the ability for overrides then

**Comment by Frontear** [comment]
Different change has been made, idea courtesy of @itslychee. There are now 5 distinct outputs that are joined in the module system to make it work identically. Normal source repo is still given, and technically the module could've worked fine without any changes, but I opted to update the module as well to use smaller outputs for a reduced closure size.

**Comment by itslychee** [comment]
Probably should split the package and module changes of the last commit into its own logically separated commits

**Comment by Frontear** [comment]
> Probably should split the package and module changes of the last commit into its own logically separated commits.
> 
> Something like
> 
> `stevenblack-blocklist: separate outputs, fix module` ->
> 
> * `stevenblack-blocklist: split host files into their own outputs`
> * `nixos/stevenblack:  allow usage of specific outputs`

Done!

**Comment by Frontear** [comment]
force push because upstream rev bump

**Comment by Frontear** [comment]
Switching to draft because smaller fixes are incubating.

File: nixos/modules/config/stevenblack.nix
Comment (Guanran928) [inline nixos/modules/config/stevenblack.nix:14]:
```suggestion
  activatedHosts = [ pkg.ads ]
    ++ optional (elem "fakenews" cfg.block) pkg.fakenews
    ++ optional (elem "gambling" cfg.block) pkg.gambling
    ++ optional (elem "porn" cfg.block) pkg.porn
    ++ optional (elem "social" cfg.block) pkg.social;
```

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (Guanran928) [inline pkgs/by-name/st/stevenblack-blocklist/package.nix:47]: Maybe `install -Dm644`?

**Review by Guanran928 — COMMENTED** [review]
Diff looks great, thanks!

File: nixos/modules/config/stevenblack.nix
Comment (Frontear) [reply nixos/modules/config/stevenblack.nix:14]: This part is changing because order no longer matters (leftover remnant from old configuration).

This is essentially the new change (largely courtesy of @Gerg-L):

```nix
options.networking.stevenblack = {
  # other options same
  package = mkPackageOption pkgs "stevenblack-blocklist" {};
};
networking.hostFiles = map (lib.getOutput x cfg.package) (cfg.package.ads ++ cfg.block)
```

Comment (JohnRTitor) [reply nixos/modules/config/stevenblack.nix:14]: Generally in Nixpkgs, people (not so much me) have been moving away from `lib.optional`, because it's apparently confusing to newcomers and easy to make mistakes with. (If you accidentally do `lib.optional cond [` in `buildInputs` you end up with a nested array which is omitted from `buildInputs`.)

Comment (Frontear) [reply nixos/modules/config/stevenblack.nix]: Resolved in latest push

**Comment by Frontear** [comment]
Done some testing, things all look a lot better overall, so I will go ahead and un-draft this

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (itslychee) [inline pkgs/by-name/st/stevenblack-blocklist/package.nix:40]: I've been wondering if it'd make more sense to just not do this and dedicate `$out` for just collecting all of the `outputs` into one place instead of just copying the repo's content over, not really sure.

**Review by itslychee — CHANGES_REQUESTED** [review]

File: nixos/modules/config/stevenblack.nix
Comment (Frontear) [reply nixos/modules/config/stevenblack.nix:43]: Afaik this wouldnt work with the types decl for the module. I honestly didn't test this specifically though so I could totally be wrong

Comment (Guanran928) [reply nixos/modules/config/stevenblack.nix:43]: Maybe like this?
```nix
block = mkOption {
  ...
  apply = x: [ "ads" x ];
}
```
Not sure if it works...

File: nixos/modules/config/stevenblack.nix
Comment (Guanran928) [inline nixos/modules/config/stevenblack.nix:18]:
```suggestion
  inherit (lib)
    getOutput
    maintainers
    mkEnableOption
    mkIf
    mkPackageOption
    mkOption
    types
    ;
```

Comment (Guanran928) [reply nixos/modules/config/stevenblack.nix:18]: Oh god, the diff looks ugly...

Remove `elem` and `optionals`

Comment (Guanran928) [reply nixos/modules/config/stevenblack.nix:18]: Also sorting them (via Vim's `:sort`) is recommended

Comment (Frontear) [reply nixos/modules/config/stevenblack.nix:18]: It does look kinda ugly 😭 but this is how nixfmt wanted it. Will sort via nvim sort tho

File: nixos/modules/config/stevenblack.nix
Comment (Frontear) [reply nixos/modules/config/stevenblack.nix:43]: That could work, though I don't see how that would be any better than doing what I've currently done.

File: pkgs/by-name/st/stevenblack-blocklist/package.nix
Comment (Frontear) [reply pkgs/by-name/st/stevenblack-blocklist/package.nix:40]: Took your advice from this, it brings all alternates and base hosts file into scope, but leaves behind all the extraneous, non-essential things that aren't likely of interest to users.

**Comment by Frontear** [comment]
Another upstream bump happened, so another rebase has been pushed

**Review by JohnRTitor — APPROVED** [review]
LGTM

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 321649` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>stevenblack-blocklist</li>
    <li>stevenblack-blocklist.ads</li>
    <li>stevenblack-blocklist.fakenews</li>
    <li>stevenblack-blocklist.gambling</li>
    <li>stevenblack-blocklist.porn</li>
    <li>stevenblack-blocklist.social</li>
  </ul>
</details>

**Comment by JohnRTitor** [comment]
Before I merge this:

```log
┃ error: builder for '/nix/store/dmr7w0cqxdj0xbz1l5fm5wm124c2jc09-review-shell.drv' failed with exit code 127;
┃        last 3 log lines:
┃        > /nix/store/1jik86ikk79rlkhrnwmyksfzcnhgaa6l-stevenblack-blocklist-3.14.84-ads: line 15: 127.0.0.1: command not found
┃        > /nix/store/d3dzfy4amjl826fb8j00qp1d9887h7hm-stdenv-linux/setup: line 692: pop_var_context: head of shell_variables not a function context
┃        > /nix/store/d3dzfy4amjl826fb8j00qp1d9887h7hm-stdenv-linux/setup: line 699: pop_var_context: head of shell_variables not a function context
┃        For full logs, run 'nix log /nix/store/dmr7w0cqxdj0xbz1l5fm5wm124c2jc09-review-shell.drv'.
```

looks like it's related to the `ads` output.

**Comment by Frontear** [comment]
> Before I merge this:
> 
> ```
> ┃ error: builder for '/nix/store/dmr7w0cqxdj0xbz1l5fm5wm124c2jc09-review-shell.drv' failed with exit code 127;
> ┃        last 3 log lines:
> ┃        > /nix/store/1jik86ikk79rlkhrnwmyksfzcnhgaa6l-stevenblack-blocklist-3.14.84-ads: line 15: 127.0.0.1: command not found
> ┃        > /nix/store/d3dzfy4amjl826fb8j00qp1d9887h7hm-stdenv-linux/setup: line 692: pop_var_context: head of shell_variables not a function context
> ┃        > /nix/store/d3dzfy4amjl826fb8j00qp1d9887h7hm-stdenv-linux/setup: line 699: pop_var_context: head of shell_variables not a function context
> ┃        For full logs, run 'nix log /nix/store/dmr7w0cqxdj0xbz1l5fm5wm124c2jc09-review-shell.drv'.
> ```
> 
> looks like it's related to the `ads` output.

Can you tell me what command you ran to get this output? Standard `nix-build -A pkgs.stevenblack-blocklist` does not reproduce this error

**Comment by Frontear** [comment]
@JohnRTitor follow-up on issue, its related to a assumption in nixpkgs-review trying to source the output: https://github.com/Mic92/nixpkgs-review/issues/407

**Comment by Frontear** [comment]
Fixed issue with stdenv sourcing the output as a shell file.

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 321649` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>stevenblack-blocklist</li>
    <li>stevenblack-blocklist.ads</li>
    <li>stevenblack-blocklist.fakenews</li>
    <li>stevenblack-blocklist.gambling</li>
    <li>stevenblack-blocklist.porn</li>
    <li>stevenblack-blocklist.social</li>
  </ul>
</details>

**Review by JohnRTitor — APPROVED** [review]
`nixpkgs-review` is fine, and built the NixOS manual with 

```bash
nix build github:nixos/nixpkgs/refs/pull/321649/merge#htmlDocs.nixosManual.x86_64-linux --print-out-paths
```

image

Let's delay this no longer and merge this when eval passes.

**Pull request merged**
